examples/research_projects/vqgan-clip/VQGAN_CLIP.py

```diff
@@ -0,0 +1,269 @@
+import os
+from glob import glob
+
+import torch
+import torchvision
+from PIL import Image
+from torch import nn
+
+import imageio
+import wandb
+from img_processing import custom_to_pil, loop_post_process, preprocess, preprocess_vqgan
+from loaders import load_vqgan
+from transformers import CLIPModel, CLIPTokenizerFast
+from utils import get_device, get_timestamp, show_pil
+
+
+class ProcessorGradientFlow:
+    """
+    This wraps the huggingface CLIP processor to allow backprop through the image processing step.
+    The original processor forces conversion to PIL images, which is faster for image processing but breaks gradient flow.
+    We call the original processor to get the text embeddings, but use our own image processing to keep images as torch tensors.
+    """
+
+    def __init__(self, device: str = "cpu", clip_model: str = "openai/clip-vit-large-patch14") -> None:
+        self.device = device
+        self.tokenizer = CLIPTokenizerFast.from_pretrained(clip_model)
```

Comment:
If using the fast tokenizers, you'll need to add the `tokenizers` library as a requirement

Reply:
Done